src/rpc.rs

```diff
@@ -901,6 +865,116 @@ impl Rpc {
+    // === tick() helpers ===
+
+    fn tick_put_queries(&mut self) -> Vec<(Id, Option<PutError>)> {
+        let mut done_put_queries = Vec::with_capacity(self.put_queries.len());
+
+        for (id, query) in self.put_queries.iter_mut() {
+            match query.tick(&self.socket) {
+                Ok(done) => {
+                    if done {
+                        done_put_queries.push((*id, None));
+                    }
+                }
+                Err(error) => done_put_queries.push((*id, Some(error))),
+            };
+        }
+
+        done_put_queries
+    }
+
+    fn tick_get_queries(&mut self) -> (Vec<(Id, Box<[Node]>)>, bool) {
+        let self_id = *self.id();
+        let responders_based_dht_size_estimate = self.responders_based_dht_size_estimate();
+        let average_subnets = self.average_subnets();
+
+        let mut done_get_queries = Vec::with_capacity(self.iterative_queries.len());
+        let mut finished_self_findnode = false;
+
+        for (id, query) in self.iterative_queries.iter_mut() {
+            if !query.tick(&mut self.socket) {
+                continue;
+            }
+
+            let closest_nodes = if let RequestTypeSpecific::FindNode(_) = query.request.request_type
+            {
+                finished_self_findnode = *id == self_id;
+
+                query
+                    .closest()
+                    .nodes()
+                    .iter()
+                    .take(MAX_BUCKET_SIZE_K)
+                    .cloned()
+                    .collect::<Box<[_]>>()
+            } else {
+                query
+                    .responders()
+                    .take_until_secure(responders_based_dht_size_estimate, average_subnets)
+                    .to_vec()
+                    .into_boxed_slice()
+            };
+
+            done_get_queries.push((*id, closest_nodes));
+        }
+
+        (done_get_queries, finished_self_findnode)
+    }
+
+    fn cleanup_done_queries(
+        &mut self,
+        done_get: &[(Id, Box<[Node]>)],
+        done_put: &mut Vec<(Id, Option<PutError>)>,
+    ) {
+        // Has to happen _before_ `self.socket.recv_from()`.
+        for (id, closest_nodes) in done_get {
+            if let Some(query) = self.iterative_queries.remove(id) {
+                self.update_address_votes_from_iterative_query(&query);
+                self.cache_iterative_query(&query, closest_nodes);
+
+                // Only for get queries, not find node.
+                if !matches!(query.request.request_type, RequestTypeSpecific::FindNode(_)) {
+                    if let Some(put_query) = self.put_queries.get_mut(id) {
+                        if !put_query.started() {
+                            if let Err(error) = put_query.start(&mut self.socket, closest_nodes) {
+                                done_put.push((*id, Some(error)))
+                            }
+                        }
+                    }
+                }
+            };
+        }
```

Comment:
This section has a lot of nesting which is hard to read/reason with. Can we do something like this to avoid deepness?

```suggestion
        for (id, closest_nodes) in done_get {
            let query = match self.iterative_queries.remove(id) {
                Some(query) => query,
                None => continue,
            };
            self.update_address_votes_from_iterative_query(&query);
            self.cache_iterative_query(&query, closest_nodes);

            // Only for get queries, not find node.
            if matches!(query.request.request_type, RequestTypeSpecific::FindNode(_)) {
                continue;
            }
            if let Some(put_query) = self.put_queries.get_mut(id) {
                if !put_query.started() {
                    if let Err(error) = put_query.start(&mut self.socket, closest_nodes) {
                        done_put.push((*id, Some(error)))
                    }
                }
            }
        }
```